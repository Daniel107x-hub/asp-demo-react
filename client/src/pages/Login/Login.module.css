form {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: left;
}

input {
    margin-bottom: 2rem;
}

.login{
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}