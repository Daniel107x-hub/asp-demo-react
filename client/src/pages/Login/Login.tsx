import React, {useState} from 'react'
import styles from './Login.module.css'
import Card from '../../components/Card/Card'
import { getUser, login } from '../../services/User/UserService'
import { Link, Navigate, useNavigate } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { setAuthenticated } from '../../app/features/Auth/authSlice'
import { setUsername } from '../../app/features/User/userSlice'; 
import { RootState } from '../../app/store'
import { toast } from 'react-toastify'

const Login = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const isAuthenticated = useSelector((state:RootState) => state.auth.isAuthenticated);
  const handleLogin = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    login(email, password)
    .then(async response => {
      if(response.status === 200) {
        dispatch(setAuthenticated(true));
        const userData = await getUser();
        const { userName } = userData.data;
        dispatch(setUsername(userName));
        return navigate('/todo');
      }
    })
    .catch(error => {
      toast.error('Unable to login with the provided credentials :(');
    });
  }
  if(isAuthenticated) return <Navigate to={'/todo'}/>
  
  const canLogin = email && password;
  return (
    <section className={styles.login}>
      <Card className={styles.formContainer}>
        <h1>Login</h1>
        <form>
          <label htmlFor="email">Email</label>
          <input type="email" name="email" id="email" value={email} onChange={e => setEmail(e.currentTarget.value)}/>
          <label htmlFor="password">Password</label>
          <input type="password" name="password" id="password" value={password} onChange={e => setPassword(e.currentTarget.value)} autoComplete='on password'/>
          <button type="submit" onClick={handleLogin} disabled={!canLogin}>Submit</button>
        </form>
        <Link to="/register">Not a user?</Link>
      </Card>
    </section>
  )
}

export default Login;

//TODO: In case the user doesn't have permission to view a page, backend will return unauthorized. Return error page on client for not authorized and a link to login again or go back