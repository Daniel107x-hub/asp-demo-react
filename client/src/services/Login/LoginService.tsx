import axios from "axios";

const login = (email: String, password: String) => {
    return axios.post('https://localhost:5105/login', { email, password });
    // return new Promise((resolve, reject) => {
    //     setTimeout(() => {
    //     if (email === 'admin' && password === 'admin') {
    //         resolve({ email, password });
    //     } else {
    //         reject('Invalid email or password');
    //     }
    //     }, 1000);
    // });
}

export {login};