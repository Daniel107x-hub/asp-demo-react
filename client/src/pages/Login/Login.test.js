import Login from "./Login"
import { renderWithProviders } from "../../utils/test-utils"
import { fireEvent, screen } from "@testing-library/react";
import { v4 as uuidv4 } from 'uuid';

describe('It loads login screen', () => {
    it('Doesn\'t break when loading', () => {
        renderWithProviders(<Login/>);
    })

    it('Matches DOM Snapshot', () => {
        const domTree = renderWithProviders(<Login/>).asFragment();
        expect(domTree).toMatchSnapshot();
    })

    it('Loads login title', () => {
        renderWithProviders(<Login/>);
        const title = screen.getByText('Login');
        expect(title).toBeInTheDocument();
    })

    it('Loads input fields empty', () => {
        renderWithProviders(<Login/>);
        const emailField = screen.getByLabelText('Email');
        expect(emailField.value).toBe('');
        const passwordField = screen.getByLabelText('Password');
        expect(passwordField.value).toBe('');
    })

    it('Submit button is disabled when fields are empty', () => {
        renderWithProviders(<Login/>);
        const emailField = screen.getByLabelText('Email');
        fireEvent.change(emailField, {target: {value: ''}});
        expect(emailField.value).toBe('');
        const passwordField = screen.getByLabelText('Password');
        fireEvent.change(passwordField, {target: {value: ''}});
        expect(passwordField.value).toBe('');
        const submitButton = screen.getByText('Submit');
        expect(submitButton).toBeDisabled();
    })

    it('Enables submit button if fields are not empty', () => {
        renderWithProviders(<Login/>);
        const randomString = uuidv4();
        const emailField = screen.getByLabelText('Email');
        fireEvent.change(emailField, {target: {value: randomString}});
        expect(emailField.value).toBe(randomString);
        const passwordField = screen.getByLabelText('Password');
        fireEvent.change(passwordField, {target: {value: randomString}});
        expect(passwordField.value).toBe(randomString);
        const submitButton = screen.getByText('Submit');
        expect(submitButton).toBeEnabled();
    })
})