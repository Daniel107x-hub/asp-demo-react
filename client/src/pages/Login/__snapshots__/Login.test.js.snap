// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It loads login screen Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <section
    class="login"
  >
    <div
      class="card formContainer"
    >
      <h1>
        Login
      </h1>
      <form>
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          type="email"
          value=""
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="on password"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <button
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
      <a
        href="/register"
      >
        Not a user?
      </a>
    </div>
  </section>
</DocumentFragment>
`;
